### DISCUSSION TITLE
how to specify a digest hash in kiali server container image when deploying via kiali-server helm chart

### DISCUSSION THREAD

**Discussion by andysworkshop** [kiali/kiali#5598] [Q&A]
https://github.com/kiali/helm-charts/blob/master/kiali-server/templates/deployment.yaml#L56

There should be an `else` on the above line. Currently if you try to use a digest you get a `:version` on the end of it.

**Comment by jmazzitelli** [comment]
@andysworkshop You did not give any details or replication procedures. You link to the "master" branch so I will assume you are trying to install the latest version using the kiali server helm chart. You did not tell us what your helm command line looks like that caused you to "get a `:version` on the end of it" nor what the resulting Deployment yaml looked like, so I cannot confirm if you found a bug or not. Please provide a complete bug report with more details rather than stating what you think is a solution to some problem. Or, better yet, open a github discussion if you have a question (such as, "How do I install the Kiali server using an image SHA digest rather than a image version tag").

As of right now, I do not see a bug here. Most likely it is user error.

Please see this in the server helm chart values files (specifically, see the comments): https://github.com/kiali/helm-charts/blob/master/kiali-server/values.yaml#L44
```
  image_digest: "" # use "sha256" if image_version is a sha256 hash (do NOT prefix this value with a "@")
```
and this a few lines down on line 48:
```
  image_version: ${HELM_IMAGE_TAG} # version like "v1.39" (see: https://quay.io/repository/kiali/kiali?tab=tags) or a digest hash
```

Here is the github enhancement request that asked for support for disconnected environments/support for hash codes: https://github.com/kiali/kiali/issues/4392

Here is the PR that introduced this support to the server helm chart: https://github.com/kiali/helm-charts/pull/108

See the test procedures in that PR for details on how to use this.

In addition to mentioning this in the server helm chart values file, we document this image_digest/image_version feature in the Operator docs for those that use the recommended install mechanism (the operator). See:

- https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.deployment.image_digest
- https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.deployment.image_version

In short, you must use the `deployment.image_digest` setting in conjunction with the `deployment.image_version` setting. `deployment.image_version` is simply the digest hash code and `deployment.image_digest` is the type of digest (most likely you want `sha256` here).

To confirm this works, use `helm template` and pass those two settings correctly, then see the container image is correctly set. For example, try this:

```
helm template --set deployment.image_digest="sha256" --set deployment.image_version="abcdef12345" kiali-server kiali/kiali-server
```

In the resultant output, look at the container image and see this:

```yaml
    spec:
      serviceAccountName: kiali
      containers:
      - image: "quay.io/kiali/kiali@sha256:abcdef12345"
```

As of right now, I will remove the bug label on this issue, as it appears to be working as intended. As stated above, the original description of this bug report did not indicate what helm command line was used and what the helm values were, so I cannot say anything more about what the problem might have been, other than to suggest it was user error.

**Comment by andysworkshop** [comment]
My mistake, apologies to you.

**Comment by jmazzitelli** [comment]
_converted to discussion and changed to a better descriptive title_

**Comment by jmazzitelli** [comment]
docs issue created that should help alleviate the confusion when it is completed : https://github.com/kiali/kiali/issues/5622

**Reply by andysworkshop** [reply]
> docs issue created that should help alleviate the confusion when it is completed : #5622

You're too kind. It really was my fault, the product of a very long day at the keyboard.

**Reply by jmazzitelli** [reply]
Wasn't just you - just today someone created yet another PR that did virtually the same thing as you proposed a couple weeks ago - so the confusion is real :)
